### ISSUE TITLE
[HOLD for payment 2023-03-13] [$1000] Whitespace above block quotes is stripped away

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#14941]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Compose a message like this:
```
I am writing a message and it includes a quote:

> This is a quote – notice that there is a blank line above it.

See how I've typed even spacing above and below that block quote?
```
2. Post the message
3. Edit the message you posted
## Expected Result:
There should be a blank line (and an even amount of space) above and below the block quote. The same message you sent in step 1 should be shown in the editor.

## Actual Result:
There's no blank line above the block quote.
Edit message shown as 
```
I am writing a message and it includes a quote:
> This is a quote – notice that there is a blank line above it.

See how I've typed even spacing above and below that block quote?
```
## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.67-4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/217424113-2b3e51e3-c0f0-468d-868c-64cd9edcb8c6.mp4

<img width="787" alt="image (2)" src="https://user-images.githubusercontent.com/43996225/217424169-dba52cb8-8970-4824-983f-885329be7575.png">
<img width="836" alt="image (1)" src="https://user-images.githubusercontent.com/43996225/217424172-1604ba28-328b-4416-9b95-dbeed88d6163.png">

**Expensify/Expensify Issue URL:**
**Issue reported by:** @roryabraham 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1675803452308849

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0162b44ca1f9c03f85</li>
        <li>Upwork Job ID: 1623397585262231552</li>
        <li>Last Price Increase: 2023-02-15</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @abekkala (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by thienlnam** [comment]
Self assigning this, will be related to the changes that we need to make in https://github.com/Expensify/App/issues/14872

**Comment by abekkala** [comment]
@thienlnam is this something that **won't** require an 'external' label for a contributor to work on and there would also be no reporting pay out since @roryabraham reported it?
I'm just making sure that an Upwork job is not needed for this particular one.

**Comment by thienlnam** [comment]
Thanks for asking, let's hold off on adding any labels right now - going to wait and see if we can fix this as part of the live markdown preview project

**Comment by roryabraham** [comment]
Just FWIW @thienlnam I feel like this could be handled separately from https://github.com/Expensify/App/issues/14872 because ultimately it's just a change inside Expensimark, which likely won't be getting changed much (at all?) by https://github.com/Expensify/App/issues/14872

**Comment by thienlnam** [comment]
Yeah good point, I'll just make this external and look over it as CME

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~0162b44ca1f9c03f85

**Comment by MelvinBot** [comment]
Current assignee @abekkala is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @rushatgabhane (`External`)

**Comment by MelvinBot** [comment]
Current assignee @thienlnam is eligible for the External assigner, not assigning anyone new.

**Comment by tsvetomir-marinov-dreamix** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Writing whitespace above block quotes is stripped away

### What is the root cause of that problem?
Looking at ExpensiMark.js seems like this is the expected behavior.

```
{
    name: 'quote',

    // We also want to capture a blank line before or after the quote so that we do not add extra spaces.
    // block quotes naturally appear on their own line. Blockquotes should not appear in code fences or
    // inline code blocks. A single prepending space should be stripped if it exists
    process: (textToProcess, replacement) => {
        const regex = new RegExp(
            /\n?^&gt; ?(?![^<]*(?:<\/pre>|<\/code>))([^\v\n\r]+)\n?/gm,
        );
        return this.modifyTextForQuote(regex, textToProcess, replacement);
    },
    replacement: g1 => `<blockquote>${g1}</blockquote>`,
},
```
            

### What changes do you think we should make in order to solve the problem?
If we comment out this functionality the problem is resolved.

Another solution is to change the regex to capture the lines only inside quotes. Example:

```
>
> First Line
>
```

**Comment by abdulrahuman5196** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
New line before a block quote is removed automatically
Note: We won't be able to repro the issue in native iOS/android.


### What is the root cause of that problem?
We have been intentionally removing a new line before a blockquote. For example 
```
'<br/><blockquote>' is changed to '<blockquote>' 
```
here https://github.com/Expensify/expensify-common/blob/5cdfe388a759e063bd86d00440a4bf277b89a3c6/lib/ExpensiMark.js#L186-L190
This change was introduced to fix this issue - https://github.com/Expensify/App/issues/9379 long time back.

```
As the regex replacement is present in the ExpensiMark.js we are removing the <br/> even before send to server for AddComment. 
But the problem is <blockquote> is behaving differently in web and differently in native iOS/android. 
In native iOS/android it automatically seems to add a new line before <blockquote> while displaying whereas in web/mweb it doesn't happen.
```

### What changes do you think we should make in order to solve the problem?

We should remove the rule to replace br before blockquote from the ExpensiMark.js. This will fix the current issue.

But it will regress the issue #9379 it originally tried to solve. So we need to apply the regex replacement before displaying the HTML in the view for native iOS/android alone to fix the original issue.

One other reason i prefer to do this change other the alternative listed below is, if we add the issue message above in a chat and try to edit the message in native iOS/android, we would find a new line missing in the edit compose window. But it would change to normal when you save it.

We could fix both issues by the above fix.

### What alternative solutions did you explore? (Optional)
I tried to explore if we could do an upstream fix in RenderHTML, but it seems there where multiple issue where people analysed on upstream fix and concluded it has to do more with rendering of blockquote at web vs native layer than the RenderHTML package scope. Also regardless of upstream fix we would be required to remove the br removal at ExpensiMark.js as its leading to the edit issue i mentioned above.
Reference
https://github.com/Expensify/App/issues/9379
https://github.com/Expensify/App/issues/12120
https://github.com/Expensify/App/issues/12784
https://github.com/Expensify/App/issues/12120

**Comment by abekkala** [comment]
@rushatgabhane can you address the question/comment by @abdulrahuman5196  above?

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $1000

**Comment by thienlnam** [comment]
We're addressing proposals, no need to double ATM

**Comment by abdulrahuman5196** [comment]
>Can we make blockquote behave the same way on web and native??

@rushatgabhane  Not sure if that's possible. I tried to search and it seems already in the referenced issues other contributors have tried to make it behave same way and it seems they weren't able to achieve expected results. And in the referenced bugs, i can see comments on **its fine to have blockquote behaving differently between web and native**. For example this issue - https://github.com/Expensify/App/issues/12784

Reference
https://github.com/Expensify/App/issues/9379
https://github.com/Expensify/App/issues/12120
https://github.com/Expensify/App/issues/12784
https://github.com/Expensify/App/issues/12120

**Comment by tienifr** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When we intentionally add a blank line above the blockquote and send it, there's no blank line above the block quote.

### What is the root cause of that problem?
The root cause is in this line https://github.com/Expensify/expensify-common/blob/5cdfe388a759e063bd86d00440a4bf277b89a3c6/lib/ExpensiMark.js#L186-L190 where we're intentionally removing the `br` above of the `blockquote`. The intention was to fix [this bug](https://github.com/Expensify/App/issues/9379).

However it didn't fix the root cause, and cause this regression: When we intentionally add the new line, it's also stripped.

### What changes do you think we should make in order to solve the problem?

We should remove that replacement logic https://github.com/Expensify/expensify-common/blob/5cdfe388a759e063bd86d00440a4bf277b89a3c6/lib/ExpensiMark.js#L186-L190 (some other changes in the PR that introduces it will need to be removed/updated as well).

And we should fix the bug there at the root cause, by enabling the `enableExperimentalBRCollapsing` flag of react-native-render-html on native platform so that the behavior of the native platform will follow the correct HTML standard.

The documentation of that flag can be found [here](https://meliorence.github.io/react-native-render-html/api/renderhtmlprops#enableexperimentalbrcollapsing).

The flag should be added to this component https://github.com/Expensify/App/blob/1b68a5f0d320665226547400f2b5e96d56825e03/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js#L241

Please note we should only enable this on native platform, this is also mentioned in the documentation
> Recommended value is true on non-web platforms

### What alternative solutions did you explore? (Optional)
NA

**Comment by tienifr** [comment]
@rushatgabhane thanks for the review!

> we'll just need to take care of all the regressions that this might introduce

I agree we'll need to do this.

**Comment by MelvinBot** [comment]
@abekkala @rushatgabhane @thienlnam this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by abekkala** [comment]
@thienlnam do you like the proposal above or need changes?

**Comment by thienlnam** [comment]
My bad, this one slipped through the cracks while I was reviewing the proposal

That proposal looks good, 🚀

**Comment by MelvinBot** [comment]
📣 @tienifr You have been assigned to this job by @thienlnam!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by tienifr** [comment]
Hi @dangrous [the PR](https://github.com/Expensify/expensify-common/pull/503) in expensify-common is ready for review. Please help me check it. Thank you!

**Comment by dangrous** [comment]
Hey! So the PR seems to work well in the test cases, and code is simple and makes sense. A couple questions just to make sure I understand the context before I merge:
* We'll be confirming in the App whether this introduced any regressions or messed something up where we had the hack before, right?
* Why does the iOS/Android additional test work right now? Shouldn't it be messed up until we turn the flag on?
Thanks!

**Comment by tienifr** [comment]
@dangrous 
> Why does the iOS/Android additional test work right now? Shouldn't it be messed up until we turn the flag on?

Because the flag was already added in this PR https://github.com/Expensify/App/pull/14822/files and turned on for native code so the additional tests for iOS/Android are working.
cc @rushatgabhane

**Comment by dangrous** [comment]
Ah! That makes much more sense.

**Comment by tienifr** [comment]
@rushatgabhane Can you help review this PR https://github.com/Expensify/App/pull/15504? Thanks

**Comment by tienifr** [comment]
Since the PR is already merged. I'm here to propose the regression test.
# Regression Test Proposal

**Bug**: Whitespace above block quotes is stripped away

**Proposed Test Steps**:

#### For all platforms
1, Compose a message like this:
```
normal text

> quote text

normal text
```
2, Post the message
3, Edit the message you posted
4, Verify that there is a blank line (and an even amount of space) above and below the block quote. The same message you sent in step 1 is shown in the editor

#### Additional test for IOS and Android (to ensure there is no regression bug created by the PR)
1, Compose a message like this:
```
normal text
> quote text
normal text
```
2, Post the message
3, Verify that there is no extra space displayed when used non-quoted and quoted then non-quoted text



Do we :+1: or :-1:

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.78-0 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15504


If no regressions arise, payment will be issued on 2023-03-13. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@rushatgabhane / @thienlnam] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@rushatgabhane / @thienlnam] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@rushatgabhane / @thienlnam] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@tienifr] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@abekkala] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by abekkala** [comment]
@tienifr I think your proposed regression test may need some tweaking. 
Also, what is the proposed location within TestRail

@rushatgabhane @thienlnam

**Comment by abekkala** [comment]
Issue reported by: @roryabraham 🎉 **no payment for you** 😉 
Selected Proposal for fix: @tienifr **[$1000]**
C+ Review: @rushatgabhane **[$1000]**

Accepted Proposal: **Feb 22**
Merged PR: **Mar 02**
**6 Business Days**
> - Merged PR within 6 business days - 0% bonus

No regressions! 🎉

**Comment by abekkala** [comment]
Offer sent to @rushatgabhane 

@tienifr I can't seem to find you in Upwork. Can you let me know your profile name within Upwork

**Comment by abekkala** [comment]
Thank you @tienifr I've sent your offer!

**Comment by abekkala** [comment]
@tienifr payment sent and contract ended - thank you! 🎉

**Comment by abekkala** [comment]
@rushatgabhane payment sent and contract ended - thank you! 🎉

**Comment by abekkala** [comment]
We're all set here - closing!
